compiler/rustc_abi/src/layout.rs

```diff
@@ -844,8 +845,11 @@ fn univariant(
             }
             // Otherwise we just leave things alone and actually optimize the type's fields
         } else {
-            let max_field_align = fields.iter().map(|f| f.align().abi.bytes()).max().unwrap_or(1);
-            let largest_niche_size = fields
+            // To allow unsizing `&Foo<Type>` -> `&Foo<dyn Trait>`, the layout of the struct must
+            // not depend on the layout of the tail.
+            let max_field_align =
+                fields_excluding_tail.iter().map(|f| f.align().abi.bytes()).max().unwrap_or(1);
+            let largest_niche_size = fields_excluding_tail
```

Comment:
These changes are necessary to fix it but I'm not sure if they're sufficient.

If an unsizing coercion could change which niche is considered the largest or how many bytes it spans that could alter the chosen layout in `LayoutCalculator::univariant`.
I haven't constructed an example yet where that actually happens. But if you want to be conservative then disabling the alt-layout branch for unsized structs would be a safe choice.

Reply:
I found a case that still fails with this PR:

```rust
#![feature(offset_of)]

#[derive(Clone)]
struct WideptrField<T: ?Sized> {
  first: usize,
  second: usize,
  niche: NicheAtEnd,
  tail: T,
}

#[derive(Clone)]
#[repr(C)]
struct NicheAtEnd {
  arr: [u8; 7],
  b: bool,
}

type Tail = [bool; 8];

fn main() {
  let sized = Box::new(WideptrField {
    first: 0xdead,
    second: 0xbeef,
    niche: NicheAtEnd {
      arr: [0; 7],
      b: false,
    },
    tail: [false; 8] as Tail,
  });

  let unsized_value: Box<WideptrField<dyn Send>> = sized.clone();

  println!("offset in niche, sized   = {}", core::mem::offset_of!(WideptrField<Tail>, niche));
  println!("offset in niche, unsized = {}", core::mem::offset_of!(WideptrField<dyn Send>, niche));
}
```

prints

```
offset in niche, sized   = 0
offset in niche, unsized = 16
```

Reply:
I have pushed a change that fixes this case. Feel free to force-push over my commit if you already have your own.